x-pack/auditbeat/processors/sessionmd/add_session_metadata.go

```diff
@@ -0,0 +1,215 @@
+	"github.com/elastic/beats/v7/x-pack/auditbeat/processors/sessionmd/processdb"
+	"github.com/elastic/beats/v7/x-pack/auditbeat/processors/sessionmd/procfs"
+	"github.com/elastic/beats/v7/x-pack/auditbeat/processors/sessionmd/provider"
+	"github.com/elastic/beats/v7/x-pack/auditbeat/processors/sessionmd/provider/ebpf_provider"
+	cfg "github.com/elastic/elastic-agent-libs/config"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/elastic-agent-libs/mapstr"
+)
+
+const (
+	processorName = "add_session_metadata"
+	logName       = "processor." + processorName
+)
+
+func init() {
+	processors.RegisterPlugin(processorName, New)
+}
+
+type addSessionMetadata struct {
+	config   config
+	logger   *logp.Logger
+	db       *processdb.DB
+	provider provider.Provider
+}
+
+func New(cfg *cfg.C) (beat.Processor, error) {
+	c := defaultConfig()
+	if err := cfg.Unpack(&c); err != nil {
+		return nil, fmt.Errorf("fail to unpack the %v configuration: %w", processorName, err)
+	}
+
+	logger := logp.NewLogger(logName)
+
+	ctx := context.Background()
+	reader := procfs.NewProcfsReader(*logger)
+	db, err := processdb.NewDB(reader, *logger)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create DB: %w", err)
+	}
+
+	backfilledPIDs := db.ScrapeProcfs()
+	logger.Debugf("backfilled %d processes", len(backfilledPIDs))
+
+	switch c.Backend {
+	case "auto":
+		// "auto" always uses ebpf, as it's currently the only backend
+		fallthrough
+	case "ebpf":
+		p, err := ebpf_provider.NewProvider(ctx, logger, db)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create ebpf provider: %w", err)
+		}
+		return &addSessionMetadata{
+			config:   c,
+			logger:   logger,
+			db:       db,
+			provider: p,
+		}, nil
+	default:
+		return nil, fmt.Errorf("unknown backend configuration")
+	}
+}
+
+func (p *addSessionMetadata) Run(ev *beat.Event) (*beat.Event, error) {
+	_, err := ev.GetValue(p.config.PIDField)
+	if err != nil {
+		// Do not attempt to enrich events without PID; it's not a supported event
+		return ev, nil //nolint:nilerr // Running on events without PID is expected
+	}
+
+	err = p.provider.UpdateDB(ev)
+	if err != nil {
+		return ev, err
+	}
+
+	result, err := p.enrich(ev)
+	if err != nil {
+		return ev, fmt.Errorf("enriching event: %w", err)
+	}
+	return result, nil
+}
+
+func (p *addSessionMetadata) String() string {
+	return fmt.Sprintf("%v=[backend=%s, pid_field=%s, replace_fields=%t]",
+		processorName, p.config.Backend, p.config.PIDField, p.config.ReplaceFields)
+}
+
+func (p *addSessionMetadata) enrich(ev *beat.Event) (*beat.Event, error) {
+	pidIf, err := ev.GetValue(p.config.PIDField)
+	if err != nil {
+		return nil, err
+	}
+	pid, err := pidToUInt32(pidIf)
+	if err != nil {
+		return nil, fmt.Errorf("cannot parse pid field '%s': %w", p.config.PIDField, err)
+	}
+
+	fullProcess, err := p.db.GetProcess(pid)
+	if err != nil {
+		return nil, fmt.Errorf("pid %v not found in db: %w", pid, err)
+	}
+
+	result := ev.Clone()
```

Comment:
This is a very expensive operation to perform for each event. The semantics of Beat processors are not well-defined in the interface. Some of the processor clone in order create a transaction-like behavior in order to be able to undo partial changes in case of any failure.

At a minimum, the Clone operation should be performed immediately before the point at which it is necessary (so right before the first action that mutates the event (`MergeFieldsDeep`)).

An alternative approach would be to perform some pre-check on the destination map to see if the merge operation can be done safely. Then there would be no need to undo a partial merge operation so we could entirely avoid the clone. We don't need do this now, but we should plan to memory profile this before making it GA.

Reply:
I've moved the Clone to just before MergeFieldsDeep.

But I think it might be possible to remove this completely. It looks like `MergeFieldsDeep` won't return error when `underRoot` is true (other than the first precondition check), although there's a comment saying the opposite: https://github.com/elastic/elastic-agent-libs/blob/v0.8.0/mapstr/mapstr.go#L375-L380.

Do you know of any conditions where it could return error and leave the event in an inconsistant state?

But I'll leave the clone in for now, and reinvestigate if it's needed when performance testing is done before GA

Reply:
I think it can fail when a field exists in the map that is not an object. Like if the event had a `process: 123` field and you wanted to merge in `{process: {pid: 123}}`.